File: Ruby/sorting/quick_sort.rb
# Quicksort
# O(n log(n)) avg time, O(n^2) worst case time, O(1) memory

def quicksort(arr, start_idx = 0, length = arr.length)
  return arr if length <= 1

  pivot_idx = partition(arr, start_idx, length)
  left_length = pivot_idx - start_idx
  right_length = length - left_length - 1

  quicksort(arr, start_idx, left_length)
  quicksort(arr, pivot_idx  + 1, right_length)
  arr
end

def partition(arr, start_idx, length)

  pivot_idx = start_idx

  (start_idx + 1...start_idx + length).each do |i|
    if arr[i] < arr[pivot_idx]
      temp = arr[i]
      arr[i] = arr[pivot_idx + 1]
      arr[pivot_idx + 1] = arr[pivot_idx]
      arr[pivot_idx] = temp
      pivot_idx += 1
    end
  end
  pivot_idx
end

File: JS/sorting/quick_sort.js
function partition (arr, startIdx, length) {
  var pivotIdx = startIdx;

  for (var i = startIdx + 1; i < startIdx + length; i++) {
    if (arr[i] < arr[pivotIdx]) {
      var temp = arr[i];
      arr[i] = arr[pivotIdx + 1];
      arr[pivotIdx + 1] = arr[pivotIdx];
      arr[pivotIdx] = temp;
      pivotIdx++;
    }
  }
  return pivotIdx;
}

function quicksort (arr, startIdx, length) {
  startIdx = typeof startIdx !== 'undefined' ? startIdx : 0;
  length = typeof length !== 'undefined' ? length : arr.length;

  if (length <= 1) return arr;

  var pivotIdx = partition(arr, startIdx, length);
  var leftLength = pivotIdx - startIdx;
  var rightLength = length - leftLength - 1;

  quicksort(arr, startIdx, leftLength);
  quicksort(arr, pivotIdx + 1, rightLength);

  return arr;
}
